08.ls_object/lib/path.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+module LS
+  class Path
+    TYPE_TABLE = {
+      '01' => 'p',
+      '02' => 'c',
+      '04' => 'd',
+      '06' => 'b',
+      '10' => '-',
+      '12' => 'l',
+      '14' => 's'
+    }.freeze
+
+    MODE_TABLE = {
+      '0' => '---',
+      '1' => '--x',
+      '2' => '-w-',
+      '3' => '-wx',
+      '4' => 'r--',
+      '5' => 'r-x',
+      '6' => 'rw-',
+      '7' => 'rwx'
+    }.freeze
+
+    HALF_YEAR = 15_768_000
+
+    attr_reader :name
+
+    def initialize(name)
+      @name = name
+    end
+
+    def blocks
+      file_stat.blocks
+    end
+
+    def type
+      TYPE_TABLE[file_mode[0..1]]
+    end
+
+    def mode
+      user_permission = change_user_permission
+      group_permission = change_group_permission
+      other_permission = change_other_permission
+      [
+        user_permission,
+        group_permission,
+        other_permission
+      ].join
+    end
+
+    def nlink
+      file_stat.nlink.to_s
+    end
+
+    def user
+      Etc.getpwuid(file_stat.uid).name
+    end
+
+    def group
+      Etc.getgrgid(file_stat.gid).name
+    end
+
+    def size
+      file_stat.size.to_s
+    end
+
+    def mtime
+      # 更新日が半年以内かどうかによって表示を変える
+      format = Time.now - HALF_YEAR < file_stat.mtime ? '%b %e %R' : '%b %e  %Y'
+      file_stat.mtime.strftime(format)
+    end
+
+    private
+
+    def file_stat
+      File.lstat(@name)
+    end
+
+    def file_mode
+      file_mode = file_stat.mode.to_s(8)
```

Comment:
```suggestion
    def file_mode
      sleep 0.5
      file_mode = file_stat.mode.to_s(8)
```

みたいにするとめちゃくちゃ遅くなりませんか？
無駄なメソッド呼び出しがないか見直してみましょう。

Reply:
https://github.com/kayama3/ruby-practices/blob/1e585703784f36931bc12438b180ac48800334a5/08.ls_object/lib/path.rb#L30-L34

file_statとfile_modeをinitializeメソッド内でインスタンス変数にすることで、メソッド呼び出し回数を減らそうとしたのですが、このやり方はアリでしょうか。
ただこの場合、file_statとfile_modeを使わない場合（lオプションが付かない場合）にも、これらの変数の初期化が行われてしまうのが気がかりです。

Reply:
> file_statとfile_modeをinitializeメソッド内でインスタンス変数にすることで、メソッド呼び出し回数を減らそうとしたのですが、このやり方はアリでしょうか。

これぐらいならありかなーと思います。

> ただこの場合、file_statとfile_modeを使わない場合（lオプションが付かない場合）にも、これらの変数の初期化が行われてしまうので、どうなのかなーと思ってます。

こちらも別に良いと思います。
もう一つの考え方として、lオプションなしの場合はPathクラスを使わない、という考え方もあります。